rotator: Refactor AVX kernels

jdemel opened the pull request:
Fix #331

This started as some work to track down the root cause of #329 . As a first result, it fixes #331 .
Apparently, these two issues are not as closely related as previously suspected.

Besides a fix for the generic kernel, this PR includes quite a bit of refactoring for better readibility of the AVX u/a rotator kernels.

jdemel committed:
rotator: Refactor AVX kernels

Fix #331

This is ongoing work to track down the root cause of #329.

1. Make sure all input scalars are normalized -> no improvement
2. Trigger renormalization after same num_points -> no improvement

Lower tolerance values move the first point that breaks this bound
further to the front. It is not obvious why this would happen so far.

Since SSE kernels tend to break later, approx. factor 2, maybe the
generic kernels accumulates errors faster. That's just a wild guess
though.

rear1019 commented on include/volk/volk_avx_intrinsics.h:
```diff
@@ -53,6 +53,17 @@ _mm256_complexconjugatemul_ps(__m256 x, __m256 y){
+static inline __m256
+_mm256_normalize_ps(__m256 val)
+{
+  __m256 tmp1, tmp2;
```
There is no need for two temporary variables.

rear1019 commented on include/volk/volk_avx_intrinsics.h:
```diff
@@ -53,6 +53,17 @@ _mm256_complexconjugatemul_ps(__m256 x, __m256 y){
+  __m256 tmp1, tmp2;
+  tmp1 = _mm256_mul_ps(val, val);
+  tmp2 = _mm256_hadd_ps(tmp1, tmp1);
+  tmp1 = _mm256_shuffle_ps(tmp2, tmp2, 0xD8);
```
You might consider using the ``_MM_SHUFFLE`` macro to make the shuffle mask more readable. ``_MM_SHUFFLE(3, 1, 2, 0)`` should be equal to ``0xD8``.

rear1019 commented on kernels/volk/volk_32fc_s32fc_x2_rotator_32fc.h:
```diff
@@ -432,103 +432,64 @@ static inline void volk_32fc_s32fc_x2_rotator_32fc_a_avx(lv_32fc_t* outVector, c
-    inc_Val = _mm256_set_ps(lv_cimag(incr), lv_creal(incr),lv_cimag(incr), lv_creal(incr),lv_cimag(incr), lv_creal(incr),lv_cimag(incr), lv_creal(incr));
-    const unsigned int fourthPoints = num_points / 4;
+    
+    const __m256 inc_Val = _mm256_set_ps(lv_cimag(incr), lv_creal(incr),
+                                         lv_cimag(incr), lv_creal(incr),
+                                         lv_cimag(incr), lv_creal(incr),
+                                         lv_cimag(incr), lv_creal(incr));
```
``_mm256_castpd_ps(_mm256_broadcast_sd((const double*)&incr))`` (see e.g. ``volk_32fc_x2_s32f_square_dist_scalar_mult_32f_a_avx()``). On the other hand, the compiler probably figures this out.

Also applies to the unaligned version.

jdemel replied:
I didn't change the implementation here. But I like this solution because it is verbose without any assumptions. Thus, I'd prefer to leave it as is.

rear1019 reviewed (approved):
It’s up to you whether to ignore my nit-picking or not.

michaelld commented on kernels/volk/volk_32fc_s32fc_x2_rotator_32fc.h:
```diff
@@ -98,18 +98,17 @@ static inline void volk_32fc_s32fc_x2_rotator_32fc_generic(lv_32fc_t* outVector,
         (*phase) *= phase_inc;
     }
-
+    if(i){
```
Do you really mean `if(i)` here? Unless `num_points` or `ROTATOR_RELOAD` is 0, then this will always be true. Or am I misunderstanding?

jdemel replied:
At first, I thought this is unnecessary but
`if(num_points%ROTATOR_RELOAD == 0)`
the loop just in front of it sets `i = 0` but then runs zero iterations. In this case, we just normalized phase in L102.

I could remove this `if` statement (and ignore that special case) but that would be inconsistent with the other kernels. Thus, I decided to leave it as is.

jdemel replied:
@michaelld is this comment resolved? In that case, we can merge this PR.

michaelld replied:
yes ... thx for the ping ... I'll do one final review & then deal with this

jdemel committed:
rotator: Remove unnecessary variable

This commit removes an unnecessary variable in `_mm256_normalize_ps`
and uses `_MM_SHUFFLE` instead of the less verbose literal `0xD8`.

jdemel commented:
@rear1019 thanks for your comments. I've updated my PR accordingly.

michaelld reviewed (approved):
LGTM ; mostly wondering about the one place I commented.

michaelld commented:
Relevant CI passes. Comments addressed. Maintainer Approval. Merging.
